src/Microsoft.AspNetCore.Mvc.TagHelpers/InputTagHelper.cs

```diff
@@ -279,54 +279,30 @@ private void GenerateCheckBox(ModelExplorer modelExplorer, TagHelperOutput outpu
                        "type",
                        "checkbox"));
             }
-
-            // Prepare to move attributes from current element to <input type="checkbox"/> generated just below.
-            var htmlAttributes = new Dictionary<string, object>(StringComparer.OrdinalIgnoreCase);
-
-            // Perf: Avoid allocating enumerator
-            // Construct attributes correctly (first attribute wins).
-            for (var i = 0; i < output.Attributes.Count; i++)
+            var hiddenForCheckboxTag = Generator.GenerateHiddenForCheckbox(ViewContext, modelExplorer, For.Name);
+            if (hiddenForCheckboxTag != null)
             {
-                var attribute = output.Attributes[i];
-                if (!htmlAttributes.ContainsKey(attribute.Name))
+                var renderingMode =
+                    output.TagMode == TagMode.SelfClosing ? TagRenderMode.SelfClosing : TagRenderMode.StartTag;
+                hiddenForCheckboxTag.TagRenderMode = renderingMode;
+
+                if (ViewContext.FormContext.CanRenderAtEndOfForm)
                 {
-                    htmlAttributes.Add(attribute.Name, attribute.Value);
+                    ViewContext.FormContext.EndOfFormContent.Add(hiddenForCheckboxTag);
+                }
+                else
+                {
+                    output.PostElement.AppendHtml(hiddenForCheckboxTag);
                 }
             }
 
-            var checkBoxTag = Generator.GenerateCheckBox(
+            var tagBuilder = Generator.GenerateCheckBox(
```

Comment:
Different styles is all. I don't like to build methods based off of how a caller treats their return values 😕